Remove invalid user EIDs and UIDs from bid request

Pubmatic-Supriya-Patil opened the pull request:
This PR solves issue https://github.com/prebid/prebid-server/issues/3859

Pubmatic-Supriya-Patil commented:
Hi @bsardo @bretg 
Please review this PR and let me know if any changes are required.

SyntaxNode commented on errortypes/code.go:
```diff
@@ -35,6 +35,8 @@ const (
 	SecCookieDeprecationLenWarningCode
 	SecBrowsingTopicsWarningCode
+	InvalidUserEIDsWarningCode
+	InvalidUserUIDsWarningCode
```
Do these new warning codes match with PBS-Java? We try to keep in sync.

Pubmatic-Supriya-Patil replied:
Can you please help me to get those error codes. Do we have any document or PR to check new warning codes match with PBS-Java.

Pubmatic-Supriya-Patil replied:
As per [prebid-server-java/pull/3465](https://github.com/prebid/prebid-server-java/pull/3465), I understand that the same 999 warning code is being used when removing eids arrays. In the Prebid Server Go repository, 999 corresponds to UnknownErrorCode in the error codes reference. Are we intending to use UnknownErrorCode here as well?

In this PR, I introduced InvalidUserEIDsWarningCode (10013) and InvalidUserUIDsWarningCode (10014). If we are okay with a generic error code, we could use UnknownErrorCode (999). However, please confirm whether we want separate error codes for InvalidUserEIDs and InvalidUserUIDs, or if using 999 would suffice.

Kindly provide your input on this.

SyntaxNode commented on endpoints/openrtb2/sample-requests/invalid-whole/user-ext-eids-uids-id-empty.json:
```diff
@@ -1,42 +0,0 @@
-{
```
Instead of removing these tests cases, please modify them to prove the new behavior.

SyntaxNode commented on endpoints/openrtb2/auction.go:
```diff
@@ -1296,8 +1296,14 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
 	if eids != nil {
-		eidsValue := *eids
-		for eidIndex, eid := range eidsValue {
+
```
Nitpick; Please remove the leading empty line.

SyntaxNode commented on endpoints/openrtb2/auction.go:
```diff
@@ -1296,8 +1296,14 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+
+		if len(eidErrors) > 0 {
+			errL = append(errL, eidErrors...)
+		}
```
Likely fine to keep this logic here for now. We want to separate validation from fixing in the future. We can refactor this new logic along with the rest later. Thoughts @bsardo?

SyntaxNode commented on endpoints/openrtb2/auction.go:
```diff
@@ -1317,6 +1323,46 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+			validEIDs = append(validEIDs, eid)
+		} else {
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("Removed EID with empty UIDs (source: %s)", eid.Source),
```
Please keep with the established error message format. Perhaps something like:

```
request.user.ext.eids[0] (source: %s) contains only empty uids and is removed from the request
```

SyntaxNode commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5922,3 +5922,179 @@ func sortUserData(user *openrtb2.User) {
+		expectedErrorMessages []string
+	}{
+		{
+			name: "Valid EID with non-empty UID",
```
Go test names cannot contain spaces so they are replaced with a dash character. This makes it hard to locate the failing test based on its name. Recommend using simpler test names like:

- `one-eid-one-uid-valid`
- `one-eid-one-uid-empty`
- `many-mixed`
- `one-eid-many-uid-empty`

The parent test name is within scope so it's easy to locate in source. Please apply this guidance to other tests added in this PR.

SyntaxNode commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5922,3 +5922,179 @@ func sortUserData(user *openrtb2.User) {
+		t.Run(tc.name, func(t *testing.T) {
+			validEIDs, errorsList := validateEIDs(tc.input)
+
+			if len(validEIDs) != len(tc.expected) {
```
Please use testify assert statements where possible. This can be written as follows which will compare the contents instead of just the array lengths, yielding more concrete tests.

```
assert.ElementsMatch(e, tc.expected, validEIDs)
```

SyntaxNode commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5922,3 +5922,179 @@ func sortUserData(user *openrtb2.User) {
+			expectedValidUIDs: nil,
+			expectedErrors:    nil,
+		},
+	}
```
Please add a test case for a nil input.

bsardo commented:
Hi @Pubmatic-Supriya-Patil, can you please merge with master (no rebase) and resolve conflicts? Thanks!

SyntaxNode commented:
Thank you for contributing this feature Please address the comments and we will re-review.

Pubmatic-Supriya-Patil commented:
> Hi @Pubmatic-Supriya-Patil, can you please merge with master (no rebase) and resolve conflicts? Thanks!

HI @bsardo I have updated branch and addressed review comments.

VeronikaSolovei9 commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,25 +1277,73 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+		if uid.ID != "" {
+			validUIDs = append(validUIDs, uid)
+		} else {
+			uidErrors = append(uidErrors, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids.uids[%d] contains empty ids and is removed from the request", uidIndex),
+				WarningCode: errortypes.InvalidUserUIDsWarningCode,
```
I used the example of user.EIDs from the issue description: 
```json
"eids": [
			{
				"source": "uidapi.com",
				"uids": [
					{"id": ""},
					{"id": "AgAACYa0cG2jcVcM42xG"}
				]
			},
			{
				"source": "liveramp.com",
				"uids": [
					{"id": "AgAACYa0cG2jcVcM42xG"}
				]
			},
			{
				"source": "yahoo.com",
				"uids": [
					{"id": ""},
					{"id": ""}
				]
			}
		]
```
As a result I received a list of errors that looked like this: 
```
request.user.eids.uids[0] contains empty ids and is removed from the request
request.user.eids.uids[0] contains empty ids and is removed from the request
request.user.eids.uids[1] contains empty ids and is removed from the request
request.user.eids[2] (source: yahoo.com) contains only empty uids and is removed from the request
```

In order to avoid confusion and duplicates may you please pass eidIndex to `validateUIDs`:
Lines 1311 and 1312:
```Golang
for eidIndex, eid := range eids {
		validUIDs, uidErrors := validateUIDs(eid.UIDs, eidIndex)
```
and then add eidIndex to error message like this: 
Line 1338:
```Golang
Message:     fmt.Sprintf("request.user.eids[%d].uids[%d] contains empty ids and is removed from the request", eidIndex, uidIndex),
```

This will give error list with no duplicate error messages: 
```
request.user.eids[0].uids[0] contains empty ids and is removed from the request
request.user.eids[2].uids[0] contains empty ids and is removed from the request
request.user.eids[2].uids[1] contains empty ids and is removed from the request
request.user.eids[2] (source: yahoo.com) contains only empty uids and is removed from the request
```

VeronikaSolovei9 replied:
Thank you for the change!

VeronikaSolovei9 commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,25 +1277,73 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+			}
+
+			for uidIndex, uid := range eid.UIDs {
+				if uid.ID == "" {
+					return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+				}
```
At this time we will not have empty `did.ID`, correct? there is a check for this already in line 1334.

VeronikaSolovei9 replied:
Also do we need to set "clean" EIDs back to `req.User.EIDs`?

Pubmatic-Supriya-Patil replied:
Can we remove if len(eid.UIDs) == 0 { validation also from validateUser function. This is also not possible because validateEids wil return error for emtpy uids. Please provide your suggestions @VeronikaSolovei9

VeronikaSolovei9 replied:
No no, I didn't ask to delete `if len(eid.UIDs) == 0 {` - it is needed here. I only asked about if we want to add valid aids to the request. And you fixed this by adding this line: `req.User.EIDs = validEids`

Pubmatic-Supriya-Patil replied:
Okay. Thanks. Please review PR

ShriprasadM commented:
@Pubmatic-Supriya-Patil  : Please respond to this comments.

bsardo commented:
@Pubmatic-Supriya-Patil can you please merge with master? That should resolve the failing validate-merge check.

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,25 +1277,69 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+		for eidIndex, eid := range validEids {
+			if eid.Source == "" {
+				return append(errL, fmt.Errorf("request.user.eids[%d] missing required field: \"source\"", eidIndex))
+			}
 
-		for uidIndex, uid := range eid.UIDs {
-			if uid.ID == "" {
-				return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+			if len(eid.UIDs) == 0 {
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
```
As per my understanding, the logic for removing invalid IDs is specific to eids. We have retained the existing logic for checking the source. However, it is unclear whether we need to check for an empty source and remove that entry while allowing other valid eids. @VeronikaSolovei9, could you please confirm this scenario and clarify the expected behavior in this case?

VeronikaSolovei9 replied:
I will ask team about the expected behavior.

bretg replied:
My take is that EID.source is required. IMO, makes sense to remove empty sources.

bsardo replied:
The team discussed offline. We're in agreement that if `EID.source` is invalid we should simply remove the corresponding EID without failing the request. We should also add a warning message.

Pubmatic-Supriya-Patil replied:
Thanks for the update. I’ll incorporate the required changes.

VeronikaSolovei9 reviewed (approved):
Thank you for the changes, approved!

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,23 +1277,69 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+			if len(eid.UIDs) == 0 {
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+			}
+		}
```
Removed

bsardo commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
+			validEIDs, errorsList := validateEIDs(tc.input)
+
+			assert.ElementsMatch(t, tc.expected, validEIDs)
+			assert.Equal(t, tc.numErrors, len(errorsList))
+
+			// Assert error messages
+			if tc.numErrors > 0 {
+				var errorMessages []string
+				for _, err := range errorsList {
+					if warning, ok := err.(*errortypes.Warning); ok {
+						errorMessages = append(errorMessages, warning.Message)
+					}
+				}
+
+				for i, expectedMsg := range tc.expectedErrorMessages {
+					if i >= len(errorMessages) {
+						t.Errorf("Expected error message %q but got none", expectedMsg)
+					} else if expectedMsg != errorMessages[i] {
+						t.Errorf("Expected error message %q but got %q", expectedMsg, errorMessages[i])
+					}
+				}
+			}
```
I think we can simplify this test quite a bit by getting rid of `numErrors` and replacing this logic as follows:
```
assert.Equal(t, len(tc.expectedErrorMessages), len(errorsList))
for _, err := range errorsList {
	assert.Contains(t, tc.expectedErrorMessages, err.Error())
}
```

Pubmatic-Supriya-Patil replied:
Done

bsardo commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
+		t.Run(tc.name, func(t *testing.T) {
+			validUIDs, errorsList := validateUIDs(tc.input, 0)
+
+			assert.ElementsMatch(t, tc.expectedValidUIDs, validUIDs, "Valid UIDs mismatch")
+			assert.ElementsMatch(t, tc.expectedErrors, errorsList, "Errors mismatch")
```
Nitpick: you can delete the last parameter in these assert statements. `t.Run` will give the appropriate context should one of these assertions fail.

Pubmatic-Supriya-Patil replied:
Removed last parameter

bsardo commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,23 +1277,64 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
 		if eid.Source == "" {
-			return append(errL, fmt.Errorf("request.user.eids[%d] missing required field: \"source\"", eidIndex))
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d] missing required field: source", eidIndex),
```
Nitpick: can we change the error message a bit so that all of the messages in this function have similar wording?
`request.user.eids[%d] removed due to missing source`

Pubmatic-Supriya-Patil replied:
Done

bsardo commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,23 +1277,64 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+			validEIDs = append(validEIDs, eid)
+		} else {
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d] (source: %s) contains only empty uids and is removed from the request", eidIndex, eid.Source),
```
Nitpick: can we change the error message a bit so that all of the messages in this function have similar wording and so it is a bit more concise?
`request.user.eids[%d] (%s) removed due to empty uids`

Pubmatic-Supriya-Patil replied:
Done

bsardo commented on endpoints/openrtb2/auction.go:
```diff
@@ -1277,23 +1277,64 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
+			validUIDs = append(validUIDs, uid)
+		} else {
+			uidErrors = append(uidErrors, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d].uids[%d] contains empty ids and is removed from the request", eidIndex, uidIndex),
```
Nitpick: can we change the error message a bit so that all of the messages in this function have similar wording and so it is a bit more concise?
`request.user.eids[%d].uids[%d] removed due to empty ids`

Pubmatic-Supriya-Patil replied:
Done

bsardo commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
+		expectedEids []openrtb2.EID
+	}{
+		{
+			description: "Valid user with Geo accuracy",
```
Nitpick: use `name` instead of `description` and replace all test case name spaces with underscores

Pubmatic-Supriya-Patil replied:
Updated

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
+		numErrors             int
+		expectedErrorMessages []string
+	}{
+		{
```
Added test cases for given scenarios

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json:
```diff
@@ -35,6 +35,6 @@
   },
-  "expectedReturnCode": 400,
-  "expectedErrorMessage": "Invalid request: request.user.eids[0].uids[0] missing required field: \"id\"\n"
+  "expectedReturnCode": 200,
```
Done

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json:
Done

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/sample-requests/invalid-whole/user-eids-source-empty.json:
```diff
@@ -36,6 +36,6 @@
-  "expectedReturnCode": 400,
-  "expectedErrorMessage": "Invalid request: request.user.eids[0] missing required field: \"source\"\n"
+  "expectedReturnCode": 200,
+  "expectedErrorMessage": "request.user.eids[0] missing required field: source"
```
Done

Pubmatic-Supriya-Patil commented on endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json:
Done
